src/dune_rules/dep_conf_eval.ml

```diff
@@ -126,26 +162,54 @@ let named ~expander l =
     List.fold_left l ~init:([], Pform.Map.empty)
       ~f:(fun (builders, bindings) x ->
         match x with
-        | Bindings.Unnamed x -> (dep expander x :: builders, bindings)
-        | Named (name, x) ->
-          let x =
-            Action_builder.memoize ("dep " ^ name)
-              (let+ l = Action_builder.all (List.map x ~f:(dep expander)) in
-               List.concat l)
-          in
-          let bindings =
-            Pform.Map.set bindings (Var (User_var name))
-              (let+ paths = x in
-               Dune_util.Value.L.paths paths)
-          in
-          (x :: builders, bindings))
+        | Bindings.Unnamed x ->
+          (to_action_builder (dep expander x) :: builders, bindings)
+        | Named (name, x) -> (
+          let x = List.map x ~f:(dep expander) in
+          match
+            Option.List.all
+              (List.map x ~f:(function
+                | Simple x -> Some x
+                | Other _ -> None))
+          with
+          | Some x ->
+            let open Memo.Build.O in
+            let x = Memo.lazy_ (fun () -> Memo.Build.all x) in
```

Comment:
Why do we need this lazy?

Reply:
Because the result is used twice:
- once every time `%{var}` is expanded
- once for interpreting the whole `(deps ...)` field
Using `lazy_` here allows to share the result